algorithms/scaling/Ih_table.py

```diff
@@ -678,12 +678,16 @@ def setup_binner(self, unit_cell, space_group, n_resolution_bins):
         d_star_sq_min -= span * relative_tolerance
         step = (d_star_sq_max - d_star_sq_min) / n_resolution_bins
 
-        self.binner = ma.setup_binner_d_star_sq_step(
-            auto_binning=False,
-            d_max=uctbx.d_star_sq_as_d(d_star_sq_max),
-            d_min=uctbx.d_star_sq_as_d(d_star_sq_min),
-            d_star_sq_step=step,
-        )
+        try:
+            self.binner = ma.setup_binner_d_star_sq_step(
+                auto_binning=False,
+                d_max=uctbx.d_star_sq_as_d(d_star_sq_max),
+                d_min=uctbx.d_star_sq_as_d(d_star_sq_min),
+                d_star_sq_step=step,
+            )
+        # Catch the case of too few unique reflections to perform the binning.
+        except AssertionError:
+            self.binner = ma.setup_binner_d_star_sq_step()
```

Comment:
Can you explain what is the difference between this call and the one above (that has failed if we reach this point), i.e. why does this work when the one above doesn't? And what is this more guaranteed to not fail?

Reply:
The key thing is the default value for the parameter `auto_binning` is `True`.  If `auto_binning` is `False` and all reflections have equal _d_, such that `d_star_sq_min == d_star_sq_max` and `step == 0`, we fall foul of the `assert` at https://github.com/cctbx/cctbx_project/blob/565ab39cd9fbc222191526fe17ec5ca5cb692471/cctbx/miller/__init__.py#L1719.

I suppose an alternative would simply be to set `step` to have a minimum value of something greater than zero (which is what `auto_binning` is doing).  As to whether we should even get this far in considering such a pathological data set, or whether we should catch the issue further upstream, that requires more in-depth knowledge of what's going on in `dials.scale` than I possess.

Reply:
As far as I remember there's nothing fancy that should be going on here, simply binning some reflection table data using standard methods, indeed I think the main, max, step approach was used to get around potential issues with the autobinning methods when there aren't many reflections present. Also I'm surprised that this code is called within dials.scale if the data is all reflections with the same d, this ought to be caught at a higher level I would think.

Reply:
Having looked at this code more carefully, I don't think that wrapping this call to `ma.setup_binner_d_star_sq_step` in a try/except is the correct approach. If the error is because we are passing nonsense input (i.e. `d_star_sq_step=0`) then we should figure out why this is the case, and either ensure we don't pass nonsense values to `ma.setup_binner_d_star_sq_step` or raise a more appropriate error (perhaps the code should never even get to this point as @jbeilstenedmands suggests).

Reply:
Well the question of what this ought to be doing upstream is beyond my current level of understanding of the scaling algorithms, so I've just taken a crude approach to ensure `step` is always greater than zero.

Reply:
OK, it's hard to comment on how/why it got to this point without the context of the dials.scale job you were running.